### PR TITLE
fix: overflow check memory operations charge gas

### LINKED ISSUE #1372: bug: Gas calculation overflow in `exec_mcopy`
## Bug Report

- Target: memory\\_operations.cairo
- Category: Coding Mistakes
- Severity: Medium
- Likelihood: Medium
- Impact: Medium

### Description

The handler for the `mcopy` opcode uses an unsafe addition when computing the amount of gas to charge for the opcode.

In the following lines,

```
// GAS
let memory_expansion = Gas.max_memory_expansion_cost(
    memory.words_len, src, size, dst, size
);

// Any size upper than 2**128 will cause an OOG error, considering the maximum gas for a transaction.
let upper_bytes_bound = size.low + 31;
let (words, _) = unsigned_div_rem(upper_bytes_bound, 32);
let copy_gas_cost_low = words * Gas.COPY;
tempvar copy_gas_cost_high = is_not_zero(size.high) * 2 ** 128;

let evm = EVM.charge_gas(
    evm, memory_expansion.cost + copy_gas_cost_low + copy_gas_cost_high
);
```

the addition of `memory_expansion.cost` and `copy_gas_cost_high` can overflow, resulting in the incorrect amount of gas being charged.

### Impact

This issue could allow disproportionately expensive operations to be charged an insufficient amount of gas. This is also a deviation from the official EVM specification.

### Recommendations

Avoid adding up the individual gas amounts and charge them individually. Alternatively, handle the potential overflow.

### PR DESCRIPTION
**Pull request by obatirou**
Time spent on this PR:

## Pull request type

Please check the type of change your PR introduces:

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

Resolves #1372

## What is the new behavior?

Overflow checks for memeory operation before charge gas

- - -
This change is [Reviewable](https://reviewable.io/reviews/kkrt-labs/kakarot/1381)

### CODE REVIEW

File: src/kakarot/instructions/memory_operations.cairo [hunk 1]
```diff
@@ -110,12 +110,18 @@ namespace MemoryOperations {
         // Any size upper than 2**128 will cause an OOG error, considering the maximum gas for a transaction.
         let upper_bytes_bound = size.low + 31;
         let (words, _) = unsigned_div_rem(upper_bytes_bound, 32);
-        let copy_gas_cost_low = words * Gas.COPY;
-        tempvar copy_gas_cost_high = is_not_zero(size.high) * 2 ** 128;
+        let copy_gas_cost = words * Gas.COPY;
+
+        let size_high_is_zero = Helpers.is_zero(size.high);
+        local gas_to_charge: felt;
+        if (size_high_is_zero != FALSE) {
```
Comment (enitrat) [inline src/kakarot/instructions/memory_operations.cairo:117]:
```suggestion
        if (size.high == 0) {
```

I think that works?

File: src/kakarot/instructions/environmental_information.cairo [hunk 2]
```diff
@@ -393,19 +403,24 @@ namespace EnvironmentalInformation {
...
-        let copy_gas_cost_low = words * Gas.COPY;
-        tempvar copy_gas_cost_high = is_not_zero(size.high) * 2 ** 128;
+        let copy_gas_cost = words * Gas.COPY;
 
-        let memory_expansion = Gas.memory_expansion_cost_saturated(
-            memory.words_len, dest_offset, size
-        );
+        let size_high_is_zero = Helpers.is_zero(size.high);
+        local gas_to_charge: felt;
+        if (size_high_is_zero != FALSE) {
```
Comment (enitrat) [inline src/kakarot/instructions/environmental_information.cairo:416]: same

File: src/kakarot/instructions/environmental_information.cairo [hunk 3]
```diff
@@ -178,13 +178,18 @@ namespace EnvironmentalInformation {
         // Any size upper than 2**128 will cause an OOG error, considering the maximum gas for a transaction.
         let upper_bytes_bound = size.low + 31;
         let (words, _) = unsigned_div_rem(upper_bytes_bound, 32);
-        let copy_gas_cost_low = words * Gas.COPY;
-        tempvar copy_gas_cost_high = is_not_zero(size.high) * 2 ** 128;
+        let copy_gas_cost = words * Gas.COPY;
+
+        let size_high_is_zero = Helpers.is_zero(size.high);
+        local gas_to_charge: felt;
+        if (size_high_is_zero != FALSE) {
```
Comment (enitrat) [inline src/kakarot/instructions/environmental_information.cairo:185]: same

File: src/kakarot/instructions/environmental_information.cairo [hunk 4]
```diff
@@ -259,13 +264,18 @@ namespace EnvironmentalInformation {
...
```
Comment (enitrat) [inline src/kakarot/instructions/environmental_information.cairo:271]: same

File: tests/src/kakarot/instructions/test_memory_operations.py [hunk 5]
```diff
@@ -28,3 +33,73 @@ def test_should_load_a_value_from_memory_with_offset_larger_than_msize(
             cairo_run(
                 "test__exec_mload_should_load_a_value_from_memory_with_offset_larger_than_msize"
             )
+
+    class TestMcopy:
+
+        @given(
+            memory_init_state=lists(
+                integers(min_value=0, max_value=255), min_size=1, max_size=100
+            ),
```
Comment (ClementWalter) [inline tests/src/kakarot/instructions/test_memory_operations.py:42]: use `st.binary` instead

File: tests/src/kakarot/instructions/test_memory_operations.py [hunk 6]
```diff
@@ -28,3 +33,73 @@ def test_should_load_a_value_from_memory_with_offset_larger_than_msize(
...
+            ]
+            expected_memory_state = (
+                memory_init_state_expansion[:dst_offset_mcopy]
+                + segment_to_copy
+                + memory_init_state_expansion[dst_offset_mcopy + size_mcopy :]
+            )
+            words_len = (len(expected_memory_state) + 31) // 32
+            expected_memory_state = "".join(
+                [f"{byte:02x}" for byte in expected_memory_state]
+            ) + "00" * (words_len * 32 - len(expected_memory_state))
```
Comment (ClementWalter) [inline tests/src/kakarot/instructions/test_memory_operations.py:78]:
```suggestion
            expected_memory_state = memory_init_state + [0] * (
                max(src_offset_mcopy, dst_offset_mcopy)
                + size_mcopy
                - len(memory_init_state)
            )
            expected_memory_state[dst_offset_mcopy: dst_offset_mcopy + size_mcopy] =             memory_expansion[src_offset_mcopy: src_offset_mcopy + size_mcopy]
```

File: tests/src/kakarot/instructions/test_memory_operations.py [hunk 7]
```diff
@@ -28,3 +33,73 @@ def test_should_load_a_value_from_memory_with_offset_larger_than_msize(
...
+            assert memory == expected_memory_state
```
Comment (ClementWalter) [inline tests/src/kakarot/instructions/test_memory_operations.py:79]: either
```suggestion
            assert bytes.fromhex(memory) == expected_memory_state
```

or
```suggestion
            assert memory == expected_memory_state.hex()
```

File: tests/src/kakarot/instructions/test_memory_operations.py [hunk 8]
```diff
@@ -28,3 +33,73 @@ def test_should_load_a_value_from_memory_with_offset_larger_than_msize(
...
+
+        @given(
+            memory_init_state=lists(
+                integers(min_value=0, max_value=255), min_size=1, max_size=100
+            ),
```
Comment (ClementWalter) [inline tests/src/kakarot/instructions/test_memory_operations.py:84]: use `st.binary`

File: tests/src/kakarot/instructions/test_memory_operations.py (hunk 6 above)
Comment (obatirou) [reply tests/src/kakarot/instructions/test_memory_operations.py:78]: Words length calculation needs to be there as serde is serialiazing the memory by words

Comment (ClementWalter) [reply tests/src/kakarot/instructions/test_memory_operations.py:78]: you can incorporate in the previous length calculation

```
(
	                max(src_offset_mcopy, dst_offset_mcopy)
	                + size_mcopy
	                - len(memory_init_state)
	            )
```

File: src/kakarot/instructions/memory_operations.cairo [hunk 9]
```diff
@@ -5,7 +5,7 @@
 from starkware.cairo.common.alloc import alloc
 from starkware.cairo.common.bool import FALSE, TRUE
 from starkware.cairo.common.cairo_builtins import HashBuiltin, BitwiseBuiltin
-from starkware.cairo.common.uint256 import Uint256
+from starkware.cairo.common.uint256 import Uint256, felt_to_uint256, uint256_add
```
Comment (ClementWalter) [inline src/kakarot/instructions/memory_operations.cairo:8]:
```suggestion
from starkware.cairo.common.uint256 import Uint256
```

**Review by ClementWalter — APPROVED** [review]

**Review by ClementWalter — APPROVED** [review]

**Pull request merged**